[TRTLLM-6756][feat] Add Beam Search to TorchSampler

stnie opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added beam search support for improved text generation quality with multiple hypotheses exploration.
  * Introduced configuration option to disable overlap scheduler for performance tuning.
  * Enhanced sampling strategy system with expanded metadata support and multi-beam handling.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

- Added Beam search to Torch Sampler
- Performance needs to be improved
- Streaming and logprobs are supported
- Early finished beams are supported
- Top-K logprobs are not supported
- Beam search is not fully async yet

## Test Coverage

- test_beam_search.py is updated to test the newly introduced features and end to end tests
- test time for test_beam_search.py   1min05sec
## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
[TRTLLM-6756][feat] Enhance TorchSampler to support beam search sampling.

- Added BeamSearchArgs class and updated methods to handle beam search logic, including cache indirection updates and beam score management.
- Modified create_torch_sampler_args to include use_overlap_scheduler parameter.
- Updated sampling strategy to accommodate beam search requests.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
[TRTLLM-6756][chore] Update Beam Search after code refactoring.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #21928](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/21928/) [ run ] triggered by Bot. Commit: `80f31f7`

Funatiq commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -832,13 +832,12 @@ def create_torch_sampler_args(mapping: Mapping, *, max_seq_len: int,
+                             max_total_draft_tokens=max_total_draft_tokens,
+                             max_num_sequences=max_num_sequences,
+                             max_beam_width=max_beam_width,
+                             use_overlap_scheduler=use_overlap_scheduler)
```
Add a comma after the last parameter to keep the formatting.

stnie replied:
done

Funatiq commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -572,9 +580,28 @@ def is_generation_model(self) -> bool:
+        beam_scores: torch.Tensor
+        """Shape: batch_size, beam_width
+           Usage: Stores the current cumulative logprob of each active beam for faster sampling"""
+        active_requests: set[int]
```
Please choose a different name so that it is not confused with the executor's `active_requests`.

Funatiq commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -572,9 +580,28 @@ def is_generation_model(self) -> bool:
+                self.CACHE_INDIRECTION_SHAPE[:-1], device="cuda", dtype=torch.float32
+            ),
+            active_requests=set(),
+        )
```
We should only create these buffers if beam search is actually used (`max_beam_width > 1`).

Funatiq commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -872,6 +904,29 @@ def _tree_sampling_batch(
         return new_draft_tokens_host
 
+    def _prepare_beam_search(self, requests: list[LlmRequest]):
+        """Prepare the beam search for the requests
```
```suggestion
        """Prepare beam search for given requests.
```

stnie committed:
[TRTLLM-6756][chore] Fixed wrong shape, when not using beam search

Unsqueeze buffer returned from sampling to always contain the beam_width dimension

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22046](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22046/) [ run ] triggered by Bot. Commit: `990321c`

Funatiq commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -238,7 +239,7 @@ def is_generation_model(self) -> bool:
 # in LlmRequest.sampling_params are either None or single-element lists.
 # This helper method simplifies code using such params.
-def _unwrap_singleton(p: Optional[list[T]]) -> Optional[T]:
+def _unwrap_singleton(p: Optional[List[T]]) -> Optional[T]:
```
Oh wow, I didn't know that either. There seem to be a lot of cases in the codebase where we still use the deprecated aliases.

ixlmar replied:
Yes, this change had confused me for quite a while. Happy the Python docs clarify what the official way is supposed to be now.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -880,6 +915,29 @@ def _tree_sampling_batch(
 
+    def _prepare_beam_search(self, requests: list[LlmRequest]):
+        """Prepare the beam search for the requests
+        If the request is not active (not currently being processed),
```
Add empty line here, otherwise I think the short description might become the concatenation of L919 + L920 ff.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -1004,36 +1094,67 @@ def update_requests(
-            if get_draft_token_length(req) > 0:
-                req.py_num_accepted_draft_tokens = num_accepted
-                req.py_rewind_len = req.py_draft_pages_allocated - num_accepted
+            if req.sampling_config.beam_width > 1:
```
This looks very similar to L1097-1108. Consider introducing helper method for DRY.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +218,115 @@ def greedy_search_sampling_batch(
+    logits: torch.Tensor,
+    beam_width: int,
+    beam_search_args: BeamSearchArgs,
+    temperature: float = 1.0,
```
Let's remove the default for `temperature` here and also in `beam_search_sampling_batch_for_prefill`. All defaults are meant to be handled by `resolve_sampling_strategy`.

stnie replied:
done

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +218,115 @@ def greedy_search_sampling_batch(
+        )
+
+        # get logprobs of each beam
+        logprobs = torch.log(softmax)
```
`torch.nn.functional.log_softmax` mentions something about numerical issues. Is there a potential accuracy issue with beam scores calculated via `log(softmax)`?

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +218,115 @@ def greedy_search_sampling_batch(
+    logits = logits.tile(1, beam_width, 1)
+    # Initialize the logits of the newly created beams to 0
+    # to prevent the same token from being sampled multiple times.
+    logits[:, 1:] = 0
```
Start with `torch.zeros` and `logits[:, 0, :].copy_(..., non_blocking=True)` or similar.

stnie replied:
I created a new `torch.zeros` buffer and copied the `logits ` into it.

ixlmar reviewed (commented):
Mostly nit, stronger feelings only about where it touches the batched sampling, also in anticipation of future work (currently mainly #8581).

stnie commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -872,6 +904,29 @@ def _tree_sampling_batch(
+            if not request.is_finished and request.py_request_id not in self.store.active_requests:
+                self.store.active_requests.add(request.py_request_id)
+                self.store.cache_indirection[request.py_seq_slot, :, request.py_prompt_len :] = (
+                    torch.arange(
```
If we prefill the buffer in the prepare step, we can omit setting these values in each iteration.

stnie commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +218,115 @@ def greedy_search_sampling_batch(
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """
+    Sample <beam_width> tokens for prefill requests in parallel.
+    Replicate the logits <beam_width> times, as context requests do not have multiple beams yet.
```
done

stnie commented on tensorrt_llm/_torch/pyexecutor/_util.py:
done

stnie committed:
[TRTLLM-6756][feat] Update beam search in TorchSampler to cover more features

- Add metadata object to grouped_request to pass additional data, which is not part of the SamplingStrategy definition.
- Add  several buffer to TorchSampler Store for beam search features, which are only allocated when beam search is used
- Add support for beam search with streaming enabled
- Beam search no longer requires all beams to finish at the same iteration.
- gather_generation_logits can now be used together with beam search.
- Logprob generation is now possible with beam search enabled.  Top-k logprobs is not supported
- Updated test_beam_search.py to also cover TorchSampler
- General changes for formatting and readability

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22787](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22787/) [ run ] triggered by Bot. Commit: `26fac04`

stnie committed:
[TRTLLM-6756][chore] Update TorchSampler to cleanup code

- Updated create_torch_sampler_args and related methods to replace use_overlap_scheduler with disable_overlap_scheduler.
- Added missing disable_overlap_scheduler parameter to TorchSampler.Args in ad_executor.py
- Introduced BeamHistory data class to encapsulate beam search history, including tokens and logprobs.
- Refactored methods to create and finalize beam history, improving clarity and functionality.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
[TRTLLM-6756][test] Update unit tests for beam search sampling functionality

- Introduced a new test file `test_beam_search_util.py` containing a dummy model and utility functions for beam search testing.
- Refactored existing tests in `test_beam_search.py` to utilize the new utility functions, improving test organization and clarity.
- Added comprehensive tests for beam search sampling, including validation of output shapes, cache indirection, and cumulative log probabilities.
- Added new unit tests for beam search sampling, updating of beams and finalization of requests

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #23043](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23043/) [ run ] triggered by Bot. Commit: `9be3a85`

stnie committed:
[TRTLLM-6756][feat] Make beam search sampling asynchronous

- Updated beam search sampling logic to ensure asynchronous handling of finished beams and cache indirection

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
[TRTLLM-6756][chore] Updated beam search code to run with async sample_async

- Added new buffer to handle Beam Search sampling asynchronously, where possible
- split beam history creation and beam finalization. Beam finalization occurs in update_requests
- adjusted write_finish_reason to support beam search

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
[TRTLLM-6756][test] Updated test_beam_search.py to correctly test for the updated beam search in TorchSampler

- enhanced logprob testing to verify sum(logprobs) == cum_log_probs
- added testing for stop tokens

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run --disable-fail-fast

stnie committed:
[TRTLLM-6756][chore] Fixed formatting after merge

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
[TRTLLM-6756][chore] Fixed bugs to pass tests and improved beam search sampling

- Fixed several bugs, which caused non-beam search testcases to fail
- Adjusted test_torch_sampler.py to conform with changes in the TorchSampler
- Improved Beam search sampling with async torch operations

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

stnie committed:
[TRTLLM-6756][chore] Update formatting after rebase + merge

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #24453](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24453/) [ run ] completed with state `SUCCESS`. Commit: `17f00d0` 
[/LLM/main/L0_MergeRequest_PR pipeline #18450](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18450/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-6756][refactor] Refactor RequestGroupKey to use dataclass

- Converted RequestGroupKey from NamedTuple to a dataclass with frozen and kw_only attributes.
- Added __iter__ and __len__ methods for improved usability.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #24495](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24495/) [ run ] triggered by Bot. Commit: `da14529`

stnie committed:
[TRTLLM-6756][refactor] Update _handle_finish_reasons method in TorchSampler

- Modified the _handle_finish_reasons method to accept an additional parameter, finish_reasons_list, for correct handling of finish reasons in beam search.
- Updated calls to _handle_finish_reasons throughout the TorchSampler class to accommodate the new parameter.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #24603](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24603/) [ run ] triggered by Bot. Commit: `85ed1cc`

stnie committed:
[TRTLLM-6756][chore] Enhance TorchSampler with new setup_sampler_step method and fix bugs

- Introduced setup_sampler_step method to enable the setup process for disaggregated serving in beam search.
- Updated cache indirection initialization to use torch.zeros to prevent reading invalid values from cache_indirection
- Updated mtpSampler to correctly call  TorchSampler functions
- Fixed handle_finish_reasons by wrapping finish reasons in the FinishReason class.
- Adjusted max_lengths_tensor calculation to account for original prompt length.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #24784](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24784/) [ run ] triggered by Bot. Commit: `230d9c4`

stnie commented:
/bot run

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #24931](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24931/) [ run ] triggered by Bot. Commit: `b65da4b`

tensorrt-cicd commented:
[PR_Github #24931](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24931/) [ run ] completed with state `SUCCESS`. Commit: `b65da4b` 
[/LLM/main/L0_MergeRequest_PR pipeline #18831](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18831/) completed with status: 'FAILURE'

stnie commented:
/bot run

Funatiq commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -410,21 +219,33 @@ def validate_outputs(llm: LLM, input_prompts: list[list[int]],
         validate_output(output, input_prompts[output_idx], sampling_params)
 
 
+# End to end tests
```
If we want to add a comment here for separation I suggest something more visible. E.g. 

```suggestion
################################################################################
# End to end tests
################################################################################
```

Funatiq reviewed (commented):
How long do the additional tests take to run in CI?

stnie commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -718,57 +833,140 @@ def _meet_stop_token_criteria(request: LlmRequest):
+    def handle_logprobs(
+        self,
+        request: LlmRequest,
+        state: SampleState,
```
removed

stnie committed:
[TRTLLM-6756][refactor] Enhance beam search handling in TorchSampler and sampling utilities

- Introduced new functions to retrieve beam width parameters for input and output, improving clarity and modularity.
- Updated UtilsSamplingParams to include separate beam width parameters and a flag for beam search usage.
- Refactored beam search sampling logic to accommodate changes in beam width handling, ensuring compatibility with new parameters.
- Unified beam search sampling for context and generation requests
- Simplified code for beam history creation
- Adjusted test cases to reflect changes in beam width handling and improved logprob validation.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #25197](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25197/) [ run ] completed with state `ABORTED`. Commit: `14bb55e` 
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19053/" target="_blank">LLM/main/L0_MergeRequest_PR #19053</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19053//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #25344](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25344/) [ run ] completed with state `SUCCESS`. Commit: `88863c1` 
[/LLM/main/L0_MergeRequest_PR pipeline #19169](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19169/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-6756][chore] Update test cases for beam search sampling after merge

- Modified `model_kwargs` to include `sampler_type` for improved test configuration.
- Adjusted `llm_cuda_graph` fixture to remove unnecessary `sampler_type` parameter.
- Enhanced clarity in `test_torch_sampler.py` by adding comments regarding the `is_context_init_state` attribute.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #25359](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25359/) [ run ] completed with state `ABORTED`. Commit: `88863c1` 
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19179/" target="_blank">LLM/main/L0_MergeRequest_PR #19179</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19179//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -25,6 +25,7 @@
 
 import torch
 
+from tensorrt_llm.bindings.executor import FinishReason
```
The [idea of this module](https://github.com/NVIDIA/TensorRT-LLM/pull/8509/files#diff-0e9d7ec99bf212b4b5ec126c7267a8b4508f8a3c98975df7937bb6a402651a64R17-R18) was to not weave in the binding types.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +238,165 @@ def greedy_search_sampling_batch(
+    # NB: Modulo vocab size is necessary to prevent end_ids from being out of bounds (e.g. -1)
+    index = beam_search_args.end_ids.view(-1, 1, 1).expand(-1, beam_width_in, 1) % vocab_size
+    # Turn the mask into a tensor of 0s and 1s for multiplication
+    # NB: we use int32 because float(-inf) * 0 returns nan instead of 0 in the scatter_reduce_
```
`float(-inf) * 0.0 = nan`, but `float(-inf) * int(0) = 0`?

ixlmar commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -450,7 +273,7 @@ def test_beam_search_output_shapes(gather_context_logits: bool,
         return_generation_logits=gather_generation_logits,
         logprobs=return_log_probs,
-        end_id=-1,
+        end_id=999,
```
Based on the latest `sampling_utils.py`, `-1` should work, or not? Since users are also likely to pass `-1`, we should cover that in the tests.

stnie replied:
I will cover this in a follow up PR

ixlmar commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -461,15 +284,31 @@ def test_beam_search_output_shapes(gather_context_logits: bool,
         )
+    if return_log_probs and llm_cuda_graph.args.sampler_type == "TRTLLMSampler":
+        pytest.skip(
+            "Beam search on TRTLLMSampler does not correctly handle log_probs if called multiple times"
```
see comment on `test_beam_search_e2e`

ixlmar reviewed (commented):
Reviewed everything except `tensorrt_llm/_torch/pyexecutor/sampler.py` (in varying detail).

tensorrt-cicd commented:
[PR_Github #25364](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25364/) [ run ] completed with state `SUCCESS`. Commit: `7019a37` 
[/LLM/main/L0_MergeRequest_PR pipeline #19185](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19185/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #25546](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25546/) [ run ] triggered by Bot. Commit: `7019a37`

mikeiovine reviewed (commented):
Signing off on speculative decoding changes in `mtp.py`

lucaslie reviewed (approved):
approving auto-deploy changes

tensorrt-cicd commented:
[PR_Github #25701](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25701/) [ run ] triggered by Bot. Commit: `7019a37`

stnie commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -206,6 +238,165 @@ def greedy_search_sampling_batch(
+
+    # get the offsets of the end tokens in the logprobs tensor
+    # NB: Modulo vocab size is necessary to prevent end_ids from being out of bounds (e.g. -1)
+    index = beam_search_args.end_ids.view(-1, 1, 1).expand(-1, beam_width_in, 1) % vocab_size
```
I will cover this in a follow up PR

stnie commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -424,21 +233,35 @@ def validate_outputs(llm: LLM, input_prompts: list[list[int]],
         )
+    if return_log_probs and llm.args.sampler_type == "TRTLLMSampler":
+        pytest.skip(
+            "Beam search on TRTLLMSampler does not correctly handle log_probs if called multiple times"
```
I will cover this in a follow up PR

tensorrt-cicd commented:
[PR_Github #25759](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25759/) [ run ] triggered by Bot. Commit: `6b4f97e`

tensorrt-cicd commented:
[PR_Github #25863](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25863/) [ run ] completed with state `SUCCESS`. Commit: `3bfaa08` 
[/LLM/main/L0_MergeRequest_PR pipeline #19609](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19609/) completed with status: 'FAILURE'

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #25979](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25979/) [ run ] completed with state `SUCCESS`. Commit: `3bfaa08` 
[/LLM/main/L0_MergeRequest_PR pipeline #19703](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19703/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #26019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26019/) [ run ] triggered by Bot. Commit: `12af683`

tensorrt-cicd commented:
[PR_Github #25996](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25996/) [ run ] completed with state `ABORTED`. Commit: `f24a77c` 
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19719/" target="_blank">LLM/main/L0_MergeRequest_PR #19719</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19719//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #26048](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26048/) [ run ] completed with state `SUCCESS`. Commit: `12af683` 
[/LLM/main/L0_MergeRequest_PR pipeline #19772](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19772/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #26417](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26417/) [ run ] triggered by Bot. Commit: `00026bd`

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -196,6 +198,29 @@ class SampleStateWithMMResult:
+    strategies: list[Strategy]
+
+
+class RequestGroupValueWithMetadata(NamedTuple):
```
With `dataclasses`, one could deduplicate this a little via `class RequestGroupValueWithMetadata(RequestGroupValue): metadata: StrategyMetadata`.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -653,9 +784,12 @@ def __init__(self, args: Args):
         self._grouped_sampler_cls: Type[GroupedStrategySampler]
         if IS_FLASHINFER_AVAILABLE and not args.disable_flashinfer_sampling:
-            from .sampling_utils_flashinfer import FlashInferGroupedStrategySampler
+            if self._use_beam_search:  # Beam search requires SimpleGroupedStrategySampler
```
this condition starts to look a bit complex: IMHO, `if not self._use_beam_search and IS_FLASHINFER_AVAILABLE and not args.disable_flashinfer_sampling` is easier to read and more DRY

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -1034,7 +1311,290 @@ def process_draft_tokens(
+    def _create_beam_history(
+        self,
+        request: LlmRequest,
+    ) -> BeamHistory | None:
```
The doc-string does not explain when this returns `None`.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -1034,7 +1311,290 @@ def process_draft_tokens(
+        grouped_requests_with_metadata: dict[RequestGroupKey, RequestGroupValueWithMetadata] = {}
+        for key, value in grouped_requests.items():
+            match key.strategy:
+                case ("beam_search", _, _, _):
```
I think the whole metadata part (alongside `RequestGroupKey` and related types) should be best encapsulated within `GroupedStrategySampler`, consistently with how it was done for `STRATEGY_KEY_TYPE`. Also, this code should be there because currently it assumes that `key.strategy: GenericStrategyKeyType` (with "GenericStrategyKeyType" unconstrained) happens to be a `Strategy`, which is only correct for `SimpleGroupedStrategySampler` (and hence brittle). If this code were in a method of `SimpleGroupedStrategySampler`, this would be more robust (and more accessible to static analysis).
